### PR TITLE
feat: File tree

### LINKED ISSUE #61: feat: file tree UI in file panel?
Thanks so much for writing this plugin!

I was wondering what you think of showing the file panel as a file tree, which I think would give you a better idea of which files have changed and how the overall project structure has changed in a diff.

Something like:

```
/path/to/project/

Changes:
M  backend/
M    file_1.py
A    file_2.py
D    file_3.py
M  frontend/
A    components/
A      file_4.tsx
A      file_5.tsx
D    common/
D      file_6.tsx

Showing changes for:
abcdef..bcdefg
```

(obviously with the nice colors that already exist)

This also avoids the problem of having a really long line width since we currently have the file name, the additions/deletions, _and_ the file path:

```
M  example_file.hs 5, 14 src/MyApp/components/internal/
```

### PR DESCRIPTION
**Pull request by farbodsz**
Resolves #61.

Hello,

Following on from our conversation earlier, I've been working on the file tree view.

It's not finished, but I'll open this PR in case you have any feedback at this stage.

The next main step is constraining cursor position to disable selection of directory names. After this it's just a matter of making the file tree view an option and adding documentation.

## Todos

- [x] File tree listing for both working files and staged files.
- [x] Make file tree optional.
  - [x] Configuring listing style to be either `list` or `tree`.
  - [x] Toggling listing style from the file panel.
- [x] Collapsing / un-collapsing folders.
  - [x] Manually toggle folder collapse.
  - [x] Automatically un-collapse folders when cycling through files with `select_next_entry` / `select_prev_entry`.
- [x] Update documentation
- [x] Proper highlight groups for folders.
- [x] Git status for folders
- [x] Extras
  - [x] Git status options: disable altogether, enable for all directories, only on folded directories
  - [x] Compacting empty folders to preserve horizontal space (see the similar feature implemented in nvim-tree: feat: Option for grouping empty directories kyazdani42/nvim-tree.lua#247)

### CODE REVIEW

**Review by sindrets — CHANGES_REQUESTED** [review]
Thank you for working on this! You are on the right track here, but there are some implementation details that need some attention.

Like I mentioned: I might open a PR to your branch to address some of the implementation details, because they implicate involving several systems that are not exactly well documented, and I want to spare you the headache of deciphering undocumented code.

**Comment by farbodsz** [comment]
Thanks so much for the detailed feedback @sindrets .

I've spent some time last week looking at the render component system and how it might work with the file tree but didn't have much time to actually start implementing it properly. My plan is to work more on the file tree this week since I'll have more time from Monday onwards.

You made a lot of useful suggestions too so I'd like to go over them this week and can let you know if I have any questions along the way.

File: lua/diffview/views/file_tree/node.lua [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+---@return boolean
+function Node:has_children()
+  return self:count_children() == 0
+end
+
+---Returns an ordered list of children recursively, with their depths, by pre-order traversal of the
+---tree.
+---@return Node[]
+function Node:children_recursive(start_depth)
```
[the remark this answers is not in the document]
Comment (farbodsz) [reply lua/diffview/views/file_tree/node.lua:48]: This is true - I've realised this when testing my changes too and is something I'm planning to fix.

**Comment by farbodsz** [comment]
@sindrets Oh I'm sorry I didn't see your pull request! (I was never notified of it)

Will take a look at it now and merge in your work.

Thanks!

**Comment by farbodsz** [comment]
Ok the commit history looks a bit convoluted after the merge but it should be fine since we're squash merging this branch anyway.

Basically all the things I was going to add in this PR are things you also mentioned in your PR description, which is great. I'll put them here too for completeness (and in this PR's description).

- Documentation
- Git status for folders
- Compacting empty folders to preserve horizontal space (see the similar feature I implemented in nvim-tree: feat: Option for grouping empty directories kyazdani42/nvim-tree.lua#247)

**Comment by farbodsz** [comment]
Hey @sindrets 

I've merged in your changes and added git statuses for directories. I would like to add some configurable options for this, something like:

```lua
require("diffview").setup({
  ...

  file_panel = {
    ...
    listing_style = "tree",
    directory_statuses = "always"  -- One of 'never', 'only_folded' or 'always'
  },
  ...
}
```

What are your thoughts on this?

The other thing I wanted to ask you, is what you mean by "proper highlight groups" in a todo comment in `render.lua`:

```lua
local function render_file_tree_recurse(depth, comp)
  -- TODO: Proper highlight groups
  local conf = config.get_config()
  ...
```

It seems to me that the correct highlighting for directories is already implemented there, so I was wondering if this a todo comment from before highlighting was added, or whether you're thinking of adding some other highlighting?

**Comment by sindrets** [comment]
> I've [...] added git statuses for directories. I would like to add some configurable options for this [...]

@farbodsz  I like this idea! The options you suggested seem good, with `always` being the default.

> The other thing I wanted to ask you, is what you mean by "proper highlight groups" in a todo comment

It's preferable to use custom highlight groups, because this lets users and theme creators customize the look of the plugin without having to change basic highlight groups that are used for many other things. Take a look in `diffview/colors.lua`. We should at least add these groups:
- `DiffviewFolderName`: links to `Directory`
- `DiffviewFolderIcon`: links to `PreProc`

You can change the fold sign to use the `DiffviewNonText` group.

Also: could you make `flatten_dirs` enabled by default? Pretty much any project with any complexity to its directory structure is practically always gonna have a bunch of compactable folders for any revision.

**Comment by farbodsz** [comment]
Ok all done!

But before we merge, I was wondering something - in config options we have something like:

```lua 
  ...
  signs = {
    ...
    folder_closed = "-",
    folder_open = "-",
  },
```

and I assume this is so that the user can set folder icons, like how we use devicons on the files.

But if the user doesn't use devicons, then a sign next to the folder is still shown. Like this:

screenshot_20210923_190649

Having folder icons without using file devicons looks a little strange, and even if I set `folder_{closed|open}` to `""` then there's still an additional whitespace character:

screenshot_20210923_190946

It would be easy to make folder icons optional in one of the following ways:

1. Check if devicons is enabled. If so, use the folder signs in the config.
2. Only use the folder signs in the config if not empty/nil.
3. Additional option like `use_folder_signs` (I'm not a fan of this one)

What do you think?

**Comment by sindrets** [comment]
Here's what I think we should do:
- Move `folder_close`, `folder_open` to a new config key `icons`
  - It would be strange if there were inconsistencies between the conditions for when we use the symbols from `signs`. And I want people to be able to configure folds without enabling icons.
- Only use these if user has `use_icons` enabled.

**Comment by farbodsz** [comment]
@sindrets yep that sounds good to me. I'll make the changes now.

**Review by sindrets — APPROVED** [review]
Brilliant. Fantastic work. Many thanks!

I'll merge this as is because it works very well, but here are some TODO's for the future:
- Make `toggle_stage_entry` work on folders in the tree.
- Fold mappings in the tree.
- Indent guides. 
  - I personally really like indent guides in trees because it makes them so much easier to read in my opinion.

These are not high priority and I can take care of them later.

Again: many thanks for your work!

**Pull request merged**